src/Hazel.hs

```diff
@@ -163,27 +162,59 @@ data PrimTy
   | NatTy
   deriving (Eq, Show)
 
-data Type
+-- | 'PreType' is a 'Type' without a usage annotation.
+data PreType
   = PrimTy PrimTy
-  -- | Type of a bounded index.
+  -- | PreType of a bounded index.
   --
   -- `IndexTy 2` is the type of `Index 0` and `Index 1`. This type is analogous
   -- to `Fin` in that it describes bounded nats.
   | IndexTy Int
-  | LollyTy (Type, Usage) Type -- TODO(bts): probably remove Usage here
-  | TupleTy (Vector Type)
-  | TimesTy (Vector Type)
-  | WithTy (Vector Type)
-  | PlusTy (Vector Type)
+  | LollyTy (PreType, UsageDeclaration) PreType
+  | TupleTy (Vector PreType)
+  | TimesTy (Vector PreType)
+  | WithTy (Vector PreType)
+  | PlusTy (Vector PreType)
   deriving (Eq, Show)
 
-data Usage = Inexhaustible | UseOnce | Exhausted
+-- | 'Type' is a usage aware 'PreType'.
+data Type = Type
+  { _preType :: PreType
+  , _usage :: UsagesRemaining
+  } deriving (Eq, Show)
+
+-- | 'Type' / '_preType' lens
+preType :: Functor f => (PreType -> f PreType) -> Type -> f Type
+preType f (Type p u) = flip Type u <$> f p
+
+-- | 'Type' / '_usage' lens
+usage :: Functor f => (UsagesRemaining -> f UsagesRemaining) -> Type -> f Type
+usage f (Type p u) = Type p <$> f u
+
+data UsageDeclaration
+  = Contemplate
```

Comment:
It's worth noting in full hopper planned type theory unpack and let
probably will be the same, I think.

It's probably worth looking into how the resource calculi by eg Jesse Tov,
Amal Ahmed (I think this was done with Greg Morrisett during her phd ) and
a few others use the term pre types, but I'm stuck on a plane for the next
hour so I may just be rambling. :)

On Tuesday, May 31, 2016, Brian Schroeder <notifications@github.com> wrote:

> In src/Hazel.hs
> <https://github.com/hopper-lang/hazel/pull/11#discussion_r65253307>:
>
> > +-- | 'Type' is a usage aware 'PreType'.
> > +data Type = Type
> > +  { _preType :: PreType
> > +  , _usage :: UsagesRemaining
> > +  } deriving (Eq, Show)
> > +
> > +-- | 'Type' / '_preType' lens
> > +preType :: Functor f => (PreType -> f PreType) -> Type -> f Type
> > +preType f (Type p u) = flip Type u <$> f p
> > +
> > +-- | 'Type' / '_usage' lens
> > +usage :: Functor f => (UsagesRemaining -> f UsagesRemaining) -> Type -> f Type
> > +usage f (Type p u) = Type p <$> f u
> > +
> > +data UsageDeclaration
> > +  = Contemplate
>
> Irrelevant?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hopper-lang/hazel/pull/11/files/d714142c141e572fc84f0ad8cbeae806e02b047b#r65253307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAQwg8wBGngcJdPKXLGbQITFy1W3bb6ks5qHJVGgaJpZM4Iq5tN>
> .
>